pytorch_lightning/loops/base.py

```diff
@@ -0,0 +1,73 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from typing import Any, Optional
+from weakref import proxy
+
+import pytorch_lightning as pl
+
+
+class Loop(ABC):
+
+    def __init__(self):
+        self.iteration_count: int = 0
+        self.trainer: Optional['pl.Trainer'] = None
+
+    @property
+    @abstractmethod
+    def done(self) -> bool:
+        """Property indicating when loop is finished"""
+
+    def connect(self, trainer, *args, **kwargs) -> None:
+        """Connects Loop with all the necessary things like connectors and accelerators"""
+        self.trainer = proxy(trainer)
+
+    @abstractmethod
+    def reset(self) -> None:
+        pass
+
+    def run(self, *args: Any, **kwargs: Any) -> Any:
+        self.reset()
+        self.on_run_start(*args, **kwargs)
+
+        while not self.done:
+            try:
+                self.on_advance_start(*args, **kwargs)
+                self.advance(*args, **kwargs)
+                self.on_advance_end()
+                self.iteration_count = self.increment_iteration(self.iteration_count)
```

Comment:
Why have this set interface to increment the iteration counter?

This won't work well when progress tracking dataclasses are integrated since we have many counters to update at very specific points of the loop.

Reply:
okay, can remove the increment function. 
but I think we still want a native iteration counter for the loop regardless of any progress tracking handler on top.

Reply:
removed